pilot/model/validation.go

```diff
@@ -1222,3 +1226,210 @@ func ValidateProxyConfig(config *proxyconfig.ProxyConfig) (errs error) {
+					fmt.Errorf("string attribute for %q should not be empty", k))
+			}
+		case *mpb.Attributes_AttributeValue_DurationValue:
+			if val.DurationValue == nil {
+				errs = multierror.Append(errs,
+					fmt.Errorf("duration attribute for %q should not be nil", k))
+			}
+			if _, err := types.DurationFromProto(val.DurationValue); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		case *mpb.Attributes_AttributeValue_BytesValue:
+			if len(val.BytesValue) == 0 {
+				errs = multierror.Append(errs,
+					fmt.Errorf("bytes attribute for %q should not be ", k))
+			}
+		case *mpb.Attributes_AttributeValue_TimestampValue:
+			if val.TimestampValue == nil {
+				errs = multierror.Append(errs,
+					fmt.Errorf("timestamp attribute for %q should not be nil", k))
+			}
+			if _, err := types.TimestampFromProto(val.TimestampValue); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		case *mpb.Attributes_AttributeValue_StringMapValue:
+			if val.StringMapValue == nil || val.StringMapValue.Entries == nil {
+				errs = multierror.Append(errs,
+					fmt.Errorf("stringmap attribute for %q should not be nil", k))
+			}
+		}
+	}
+	return errs
+}
+
+// ValidateHTTPAPISpec checks that HTTPAPISpec is well-formed.
+func ValidateHTTPAPISpec(msg proto.Message) error {
+	in, ok := msg.(*mccpb.HTTPAPISpec)
+	if !ok {
+		return errors.New("cannot case to HTTPAPISpec")
+	}
+	var errs error
+	// top-level list of attributes is optional
+	if in.Attributes != nil {
+		if err := ValidateMixerAttributes(in.Attributes); err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	}
+	if len(in.Patterns) == 0 {
+		errs = multierror.Append(errs, errors.New("at least one pattern must be specified"))
+	}
+	for _, pattern := range in.Patterns {
+		if err := ValidateMixerAttributes(in.Attributes); err != nil {
+			errs = multierror.Append(errs, err)
+		}
+		if pattern.HttpMethod == "" {
+			errs = multierror.Append(errs, errors.New("http_method cannot be empty"))
+		}
+		switch m := pattern.Pattern.(type) {
+		case *mccpb.HTTPAPISpecPattern_UriTemplate:
+			if m.UriTemplate == "" {
+				errs = multierror.Append(errs, errors.New("uri_template cannot be empty"))
+			}
+		case *mccpb.HTTPAPISpecPattern_Regex:
+			if m.Regex == "" {
+				errs = multierror.Append(errs, errors.New("regex cannot be empty"))
+			}
+		}
+	}
+	for _, key := range in.ApiKeys {
+		switch m := key.Key.(type) {
+		case *mccpb.APIKey_Query:
+			if m.Query == "" {
+				errs = multierror.Append(errs, errors.New("query cannot be empty"))
+			}
+		case *mccpb.APIKey_Header:
+			if m.Header == "" {
+				errs = multierror.Append(errs, errors.New("header cannot be empty"))
+			}
+		case *mccpb.APIKey_Cookie:
+			if m.Cookie == "" {
+				errs = multierror.Append(errs, errors.New("cookie cannot be empty"))
+			}
+		}
+	}
+	return errs
+}
+
+// ValidateHTTPAPISpecBinding checks that HTTPAPISpecBinding is well-formed.
+func ValidateHTTPAPISpecBinding(msg proto.Message) error {
```

Comment:
note: we're entertaining the idea of using protoc-validate annotations from lyft to automate these functions. this is fine for now.